sympy/functions/elementary/hyperbolic.py

```diff
@@ -1702,6 +1750,25 @@ def inverse(self, argindex=1):
     def _eval_rewrite_as_log(self, arg, **kwargs):
         return log(1/arg + sqrt(1/arg - 1) * sqrt(1/arg + 1))
 
+    def _eval_rewrite_as_acosh(self, arg, **kwargs):
+        return sqrt(1/arg - 1)/sqrt(1 - 1/arg)*acosh(1/arg)
+
+    def _eval_rewrite_as_asinh(self, arg, **kwargs):
+        return sqrt(1/arg - 1)/sqrt(1 - 1/arg)*(S.ImaginaryUnit*asinh(S.ImaginaryUnit/arg)
+                                                + S.Pi*S.Half)
+
+    def _eval_rewrite_as_atanh(self, arg, **kwargs):
+        return (S.Pi*S.ImaginaryUnit*(1 - sqrt(arg)*sqrt(1/arg) -
+                (S.ImaginaryUnit*sqrt(-arg))/(2*sqrt(arg)) -
+                (S.ImaginaryUnit*sqrt(arg**2))/(2*sqrt(-arg**2)))
+                + sqrt(arg + 1)*sqrt(1/(arg + 1))*atanh(sqrt(1 - arg**2)))
```

Comment:
This doesn't always work:
```python
In [2]: f = asech(x)

In [3]: f.subs(x, -1-I).n()
Out[3]: 0.530637530952518 + 2.02307477394609⋅ⅈ

In [4]: f.rewrite(atanh).subs(x, -1-I).n()
Out[4]: 0.530637530952518 - 1.11851787964371⋅ⅈ
```

Reply:
On master this works because there is no rewrite:
```python
In [1]: f = asech(x)

In [2]: f.subs(x, -1-I).n()
Out[2]: 0.530637530952518 + 2.02307477394609⋅ⅈ

In [3]: f.rewrite(atanh).subs(x, -1-I).n()
Out[3]: 0.530637530952518 + 2.02307477394609⋅ⅈ
```

Reply:
OK! I have if nothing else double-checked that the equation is the same as https://functions.wolfram.com/ElementaryFunctions/ArcSech/27/02/10/01/01/0005/

Reply:
Both DLMF and mpmath define `asech(x)` as `acosh(1/x)`:
https://dlmf.nist.gov/4.37
https://mpmath.org/doc/current/functions/hyperbolic.html#asech

Their definitions of `atanh` don't seem to be equivalent though. mpmath says:
```
atanh(x) = 1/2 (log(1+x) - log(1-x)
```
DLMF says that away from the branch cuts
```
atanh(x) = 1/2 log((1+x)/(1-x))
```
These are not equivalent for most complex numbers:
```python
In [26]: atanh_mpmath = (log(1 + x) - log(1 - x))/2

In [27]: atanh_dlmf = log((1 + x)/log(1 - x))/2

In [28]: atanh_mpmath.subs(x, -1 + I)
Out[28]: 
  log(2 - ⅈ)   ⅈ⋅π
- ────────── + ───
      2         4 

In [29]: atanh_dlmf.subs(x, -1 + I)
Out[29]: 
   ⎛    ⅈ     ⎞
log⎜──────────⎟
   ⎝log(2 - ⅈ)⎠
───────────────
       2       

In [30]: atanh_mpmath.subs(x, -1 + I).n()
Out[30]: -0.402359478108525 + 1.01722196789785⋅ⅈ

In [31]: atanh_dlmf.subs(x, -1 + I).n()
Out[31]: 0.036967944745527 + 1.04675131315153⋅ⅈ
```
I'm also not sure that the definitions of `acosh` are equivalent either.

Reply:
Okay, so the problem is `subs` rather than `rewrite`:
```python
In [1]: f = asech(x)

In [2]: f.evalf(subs={x:-1-I})
Out[2]: 0.530637530952518 + 2.02307477394609⋅ⅈ

In [3]: f.rewrite(atanh).evalf(subs={x:-1-I})
Out[3]: 0.530637530952518 + 2.02307477394609⋅ⅈ
```